### DISCUSSION TITLE
DynUnet Transform BRATS 21

### DISCUSSION THREAD

**Discussion by nourmagde00** [Project-MONAI/MONAI#4073] [Q&A] [answered]
Could any one tell me the Monai Transforms for BRATS 21 Dataset for DynUnet model ?
I already used these Transforms below and get --> RuntimeError: Sizes of tensors must match except in dimension 1. Expected size 10 but got size 9 for tensor number 1 in the list.

Thanks in Advance
![FINAL-Jupyter-Notebook](https://user-images.githubusercontent.com/56939323/161640772-bfa961e5-da4d-4caf-a406-34314450fa52.png)

**Comment by Nic-Ma** [comment, marked as answer]
Hi @yiheng-wang-nv ,

Could you please help share some ideas here?

Thanks in advance.

**Reply by nourmagde00** [reply]
First of all ,thank you very much for your replying <3 
My graduation project is BRATS 21 Segmentation Using nnUNet or DynUNet model in MONAI , I want know the monai transforms for BRATS 21 .....so I searched and found the transforms  for BRATS 16 & 17 but when I trained the model I get this Error :
![image](https://user-images.githubusercontent.com/56939323/161680551-82f6788c-0768-47a1-b35c-c16fa172ffd2.png)

**Reply by yiheng-wang-nv** [reply]
Hi @nourmagde00 , could you please share the following info?
1. What is the original shape of the brats 21 dataset (for images and labels)? If the spatial shapes are different, what is the number of channels?
2. Could you please more error information, thus I'm able to see the issue comes from which transform.

**Reply by nourmagde00** [reply]
Hi @yiheng-wang-nv  ,Thank you for your replying <3....
The original shape is (240, 240, 155) ,the number of channels is 3 ,for error info this is my notebook , I'll be very thankful if you see it

**Reply by yiheng-wang-nv** [reply]
HI @nourmagde00 , the error message seems CUDA OOM error, rather than the size mis match error. I found that you set `patch_size = [128, 128, 128]`, and use this value to determine the kernel size and strides of the network, but the `roi_size` of your spatial crop transform is `[224, 224, 144]`, which is different from the `patch_size`. These two values should be the same. Could you please try to do: change the `roi_size`?

You can also refer to: https://docs.monai.io/en/stable/networks.html?highlight=dynunet#dynunet , the input size of DynUNet should meet some requirements, which are determined by `strides`.

**Reply by nourmagde00** [reply]
Thank you very much for your response <3 @yiheng-wang-nv ,
I tried applying your instruction by setting roi size with the same value as the patch size (224, 224, 144) but I still recieved another error at the end of the notebook "RuntimeError: Sizes of tensors must match except in dimension 1. Expected size 10 but got size 9 for tensor number 1 in the list."

[BRATS21   Jupyter Notebook.pdf](https://github.com/Project-MONAI/MONAI/files/8454839/BRATS21.Jupyter.Notebook.pdf)

**Reply by yiheng-wang-nv** [reply]
Hi @nourmagde00 ,

If `strides = [[1, 1, 1], [2, 2, 2], [2, 2, 2], [2, 2, 2], [2, 2, 2], [2, 2, 2]]`, the patch size should be divisible by `2 ** 5`, which is 32. Therefore, size 144 can not be used. Could you please try to set `patch_size = [128, 128, 128]` and `roi_size = [128, 128, 128]`?

In addition, that's for posting this question, I submitted a PR to enhance the doc-strings of the network (see: #4102 ), and I will also enhance [the DynUNet tutorial ](https://github.com/Project-MONAI/tutorials/blob/c5faf378e9289eb3efa8854bf33fab97e794ac78/modules/dynunet_pipeline/create_network.py#L9) via adding the patch size check for the function `get_kernels_strides`.

**Reply by nourmagde00** [reply]
Mr @yiheng-wang-nv  I really appreciate your responses and I'm very sorry for your effort with me ❤️❤️❤️,
I applied all of your instructions ,using roi-size =(128,128,118) crops important training parts from the brian. So, I had to add resized() in order to make the data with the shape (128,128,128), now I'm training my model 
I just want to ensure that we are in the right way ,Could you check these 3 photos , it's all about MRI preprocessing for my dataset BRATS 21 so, my question : Is that preprocessing is a well preprocessing for my dataset ? I just afraid from losing time in the training and build a weak model .
I'd like to thank you again and again for your cooperation ❤️

![Transform](https://user-images.githubusercontent.com/56939323/163100846-48a16bb5-f49a-4739-bf40-7a4e574d1127.png)
![TrainTransform](https://user-images.githubusercontent.com/56939323/163100813-3a03df02-8a22-4bde-9146-3ab18a58db24.png)
![valTransform](https://user-images.githubusercontent.com/56939323/163102500-df1f38de-787b-4b96-9e4d-9ac59af2b2a6.png)

**Reply by yiheng-wang-nv** [reply]
Hi @nourmagde00 , if `Resized` is used, you may need to think about how to convert the inference result back to the original shape, you may need to use `Invertd` transform. I suggest to use `RandSpatialCropd(keys=["image", "label"], roi_size=[128, 128, 128], random_size=False)`, rather than using `roi_size=[224, 224, 144]` with the resize transform, and you can remove the crop foreground transform to make sure that the size after crop will not be changed. 

You can also try to use `RandCropByPosNegLabeld`, which may peform better than `RandSpatialCropd` (it depends on the experiments, I'm not 100% sure and it is just an advice).

**Reply by nourmagde00** [reply]
Hi @yiheng-wang-nv  @Nic-Ma I hope u be well,
I get a train loss equal nan when I train my model , I tried to solve it for several dayes but it does't work yet,
Could u check my notebook please ?
[SEG NetJupyter Notebook.pdf](https://github.com/Project-MONAI/MONAI/files/8598135/SEG.NetJupyter.Notebook.pdf)
 
Thanks In Advance.
